### PR TITLE
X86 Elempack 8 AVX implementations.

### PR DESCRIPTION
**Pull request by Timen**
Major changes:

1. Added element packing 8 for x86
2. added element packing implementation for main conv layers (3x3 conv, 1x1 conv, 3x3 & 5x5 dwconv)
3. added x86 versions with element packing for the activation functions.
4. added x86 versions with element packing for misc layers (eltwise, concat etc.)

Most of what I did was taking the arm element pack 4 implementations and convert them to element packing 8. Not all ctests are passing as I'm having a bit of a tough time converting the 3x3s1 pack 4 to 3x3s1 due to the winograd implementation used which complicates things. (maybe @nihui you could give me some tips there?)

performance speedup is 2X for some nets but some also got slower probably due to still missing a lot of element packing implementations used in some networks. 

TODO:
1. Add reshape and channel shuffle element packing 8 implementations
2. Fix 3x3 pack 8 implementation
3. Unroll unroll unroll, lot of loops can still be unrolled.

### CODE REVIEW

**Comment by nihui** [comment]
I think we can sort out the finished parts first. About Winograd and the rest, put in subsequent PR
It doesn't have to be all realized at once. So I can review and merge this pr first.

**Comment by Timen** [comment]
> I think we can sort out the finished parts first. About Winograd and the rest, put in subsequent PR It doesn't have to be all realized at once. So I can review and merge this pr first.

Okay, I'll move that to a future PR, tests are passing now apart from the restyle test (even though I did run `./codeformat.sh`.

**Comment by nihui** [comment]
> > I think we can sort out the finished parts first. About Winograd and the rest, put in subsequent PR It doesn't have to be all realized at once. So I can review and merge this pr first.
> 
> Okay, I'll move that to a future PR, tests are passing now apart from the restyle test (even though I did run `./codeformat.sh`.

codeformat script needs clang-format 10.0 and astyle 3.1

You can check these style changes here
https://github.com/Timen/ncnn/compare/tijmen-conv-avx-packing...Tencent:restyled/pull-1853

**Comment by Timen** [comment]
Okay I merged the restyle pr to fix the styling. I double checked my astyle and clang-format versions and they were 3.1 and 8.0.0 respectively, however after upgrading clang-format. to 10.0.0  `./codeformat.sh` still didn't produce the correct formatting.

File: tests/testutil.h [hunk 1]
```diff
@@ -543,7 +547,11 @@ int test_layer(int typeindex, const ncnn::ParamDict& pd, const std::vector<ncnn:
         ncnn::Mat a4;
         if (opt.use_packing_layout)
         {
+#if (defined(__x86_64__) || (defined _WIN32 && !(defined __MINGW32__)))
```
[the remark this answers is not in the document]
Comment (tijmen-loomai) [reply tests/testutil.h:550]: For some reason msvc doesn’t set the __AVX__ compiler directive for the ctest files. So instead i simply detect if we are executing on a x86 platform. I can check if i can fix the avx directive.

[the remark this answers is not in the document]
Comment (nihui) [reply tests/testutil.h:550]: ok, I think you are right actually.
for test program, we shouldn't rely on compiler definitions, but the platform target one.

**Comment by bigbigzxl** [comment]
good night～

File: src/layer/x86/pooling_2x2.h [hunk 2]
```diff
@@ -15,34 +15,40 @@
...
+                                 const Option& opt)
+{
     int w = bottom_blob.w;
     int inch = bottom_blob.c;
 
     int outw = top_blob.w;
     int outh = top_blob.h;
+#if __AVX2__
+    __m256i permute_mask = _mm256_setr_epi32(0, 2, 4, 6, 1, 3, 5, 7);
+#endif // __AVX__
```
[the remark this answers is not in the document]
Comment (Timen) [reply src/layer/x86/pooling_2x2.h]: done!

File: src/layer/x86/reshape_x86.cpp [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        delete flatten;
+        flatten = 0;
+    }
+
+    return 0;
+}
+
+int Reshape_x86::forward(const Mat& bottom_blob, Mat& top_blob, const Option& opt) const
+{
+    // #if __AVX__
```
[the remark this answers is not in the document]
Comment (Timen) [reply src/layer/x86/reshape_x86.cpp]: removed, will add it back next week i think with the element packing avx support.

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1853) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1853) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: Timen<br/>:x: restyled-commits<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=1853) it.</sub>

**Comment by nihui** [comment]
Thanks !

**Comment by Timen** [comment]
No problem! I did discover an issue with the bench script where the int8 models were crashing. I wanted to fix this before the merge but you were to fast! Anyways I’ll do this after the weekend and then I’ll also fix the 3x3 winograd implementation👍

**Comment by nihui** [comment]
Don't worry about these temporary crashes, that's how git master is.  :laughing:

**Pull request merged**